--- packages/backend/node_modules/libp2p/dist/src/upgrader.js	2025-01-16 17:31:38
+++ packages/backend/patch/upgrader.js	2025-01-16 17:22:58
@@ -364,7 +364,8 @@
                 catch (err) {
                     connection.log.error('could not create new outbound stream on connection %s %a for protocols %s - %e', direction === 'inbound' ? 'from' : 'to', opts.maConn.remoteAddr, protocols, err);
                     if (muxedStream.timeline.close == null) {
-                        muxedStream.abort(err);
+                        // muxedStream.abort(err);
+                        await muxedStream.close()
                     }
                     throw err;
                 }
\ No newline at end of file
